README.md

                                          Basic Linux/Unix shell
                                                            
                                                            

I had implemented two types of bash commands:

1) Internal commands

I had implemented five Internal commands like ‘cd’, ‘echo’, ‘history’, ‘pwd’ and ‘exit’.

'cd' - Helps in changing current directory.
followed by directory path and prints error if path doesn't exists.
Flags Implemented:
-e - return -1 if directory path doesn't exists.
--help - Demonstrate how to use cd commands along with implemented flags.

'echo' - Prints the statement followed by this bsh command along with newline character.
Flags Implemented:
-n - prints the statement without newline character
-E - default implementation of echo command with implemented flags.

'history' - Prints the history of commands used in current session of terminal
Flags Implemented:
-w - stores the history of current bash session to history.txt file for permanent store.
--help - Demonstrate how to use history commands along with implemented flags.

'pwd' - Helps to get current path of file and prints error if extra inputs are provided.
Flags Implemented:
-L - default implementation of pwd is done by this flag in terminal.
--help - Demonstrate how to use pwd commands along with implemented flags.

'exit' - exits the terminal session



2) External commands
 
I had implemented five External commands like ‘ls’, ‘cat’, ‘date’, ‘rm’ and ‘mkdir’.

'ls' - Tells the name of all files/folders present in current directory sorted in ascending order. Case for ls for different directory in one command is also handled.
Flags Implemented:
-1 - Prints the name of all files/folders present in current directory sorted in ascending order in vertical lines instead of default horizontal alignmnent.
--help - Demonstrate how to use ls commands along with implemented flags.

'cat' - Concatenates the content of given numbers files and display on console. Prints error if any file is not present along with output of files presented.
Flags Implemented:
-n - Prints the number along with each line.
-E - Prints $ at the end of each line.
--help - Demonstrate how to use cat commands along with implemented flags.

'date' - Prints standard system date with console format.
Flags Implemented:
-R - prints current date in UTC time zone format.
-u - prints current UTC zone time.
--help - Demonstrate how to use date commands along with implemented flags.

'rm' - use to delete listed files as arguments and shows error if any file doesn't exists. It does not remove directory by default.
Flags Implemented:
-d - use to empty directory if present else shows error.
--help - Demonstrate how to use delete commands along with implemented flags.

'mkdir' - use to make new directory at the current pwd location.
Flags Implemented:
-v - prints success message on console whenever directory is created successfully.
--help - Demonstrate how to use delete commands along with implemented flags.


Implementation of external commands is done by fork() system call, creating child process for external bash command and replacing child code by execvp passing file path along with double pointer to specific file.
'date' -> date.c
'ls' -> ls.c
'history' -> history.c
'rm' -> rem.c
'cat' -> cat.c




Common Errors Handled:

1) Shows bash error for any wrong command except of above listed 10 implemented commands.
2) Unknown flag error.
3) Insufficient/Extra provided arguments to particular command.


Testcase and Sceenshots:

Basic test case are also provided along with their output in the form of practical runnung by screenshots. Kindly refer to them for better understanding. 
Screenshots are present in 'Screenshots' folder.

Sample inputs:

ls 

ls -l 

mkdir 

mkdir nik@1 

rm -d nik@1 

date 

date cdmv 

pwd 

cd.. 

cd .. 

pwd 

history 

history -w 

exit 

cat history.txt 

exit 


NOTE:

Kindly remember to use '@' delemiter instead of space for passing name of files or text to be echoed.
                                                  


history.txt
History:
cat
cat
hstory
history
history-

ls.c
#include<stdio.h>
#include<string.h>
#include<stdlib.h>
#include<dirent.h>
extern  int alphasort();
struct dirent **names;

char *org(char *name)
{
    char* tmp=name;
    while (*tmp!='\0')
    {
        if(*tmp=='@')
            *tmp=' ';
        tmp++;
    }
    return name;         
}

int printls(char* s,int cc)
{
    struct dirent **names;
    int i;
    i=scandir(s,&names,NULL,alphasort);
    if(i<0)
    {
        printf("ls: cannot access '%s': No such file or directory\n",s);
        return(-1);
    }
    else
    {
        for(int k=0;k<i;k++)
        {
            if(cc==0)
                printf("%s      ",names[k]->d_name);
            else
                printf("%s\n",names[k]->d_name);
            free(names[k]);
        }
        if(cc!=1)
            printf("\n");
        free(names);
    }
}
int main(int argc,char **argv)
{   
    if(argc>=2 && strcmp(argv[1],"-1")==0)
    {
        if(argc==2)
        {
            printls(".",1);
            return(0);
        }
        if(argc>2)
        {
            int i=2;
            while (argv[i]!=NULL)
            {
                printf("%s:\n",org(argv[i]));
                printls(org(argv[i]),1);
                i++;
            }
            return(0);
        }
    }

    else if(argc!=1&&strcmp(argv[1],"--help")==0)
    {
        printf("Usage: ls [OPTION]... [FILE]...\n");
        printf("List information about the FILEs (the current directory by default).\n");
        printf("Sort entries alphabetically if none of -cftuvSUX nor --sort is specified.\n\n");
        printf("  -1,             list one file per line.\n");
        printf("  --help  display this help and exit\n\n");
        printf("Exit status:\n");
        printf("  0  if OK,\n");
        printf("  1  if minor problems (e.g., cannot access subdirectory),\n");
        printf("  2  if serious trouble (e.g., cannot access command-line argument).\n");
        return(0);
    }
    else
    {
        if (argc!=1)
        {
            char* tmp=argv[1];
            if (*tmp=='-')
            {
                printf("ls: invalid option -- '%s'\n",argv[1]);
                printf("Try 'ls --help' for more information.\n");
                return(-1);
            }
        }
        
        int i=1;
        if(argc==1)
            printls(".",0);
        else
        {
            char* tmp=argv[1];
            if (*tmp=='-')
            {
                printf("ls: invalid option -- '%s'\n",argv[1]);
                printf("Try 'ls --help' for more information.\n");
                return(-1);
            }
            else
            {
                while (argv[i]!=NULL)
                {
                    printf("%s:\n",org(argv[i]));
                    printls(org(argv[i]),0);
                    i++;
                }
            }
            
        }
        
        return(0);
    }
}

cat.c
#include<stdio.h>
#include<unistd.h>
#include<string.h>
#include<sys/stat.h>
#include<sys/types.h>
#include<stdlib.h>

int count=1;
char *org(char *name)
{
    char* tmp=name;
    while (*tmp!='\0')
    {
        if(*tmp=='@')
            *tmp=' ';
        tmp++;
    }
    return name;
}

int red_num(char *file)
{
    FILE *file_ptr;
    char c;
    file_ptr=fopen(file,"r");
    if(file_ptr == NULL)
    {
      printf("cat: %s: No such file or directory\n",file);   
      return(0);             
    }
    printf("   %d ",count);
    while (1)
    {
        c = fgetc(file_ptr);
        if(c==EOF)
            break;
        else
        {
            printf("%c", c);
            if(c=='\n')
            {
                count+=1;
                printf("   %d ",count);
            }

        }
    }
    fclose(file_ptr);    
}

int red(char *file)
{
    FILE *file_ptr;
    char c;
    file_ptr=fopen(file,"r");
    if(file_ptr == NULL)
    {
      printf("cat: %s: No such file or directory\n",file);   
      return(0);             
    }
    while (1)
    {
        c = fgetc(file_ptr);
        if(c==EOF)
            break;
        else
            printf("%c", c);
    }
    fclose(file_ptr);    
}

int red_dol(char *file)
{
    FILE *file_ptr;
    char c;
    file_ptr=fopen(file,"r");
    if(file_ptr == NULL)
    {
      printf("cat: %s: No such file or directory\n",file);   
      return(0);             
    }
    while (1)
    {
        c = fgetc(file_ptr);
        if(c==EOF)
            break;
        else
        {
            if(c=='\n')
                printf("$");
            printf("%c", c);
        }
    }
    fclose(file_ptr);    
}

int main(int argc,char **argv)
{   
    if(strcmp(argv[1],"-n")==0)
    {
        int i=2;
        int flag=0;
        while(argv[i]!=NULL)
        {
            flag=1;
            red_num(org(argv[i]));
            i+=1;
        }
        count=0;
        if(flag==0)
        {
            while(1)
            {
                char *st;
                fgets(st,10000,stdin);
                printf("%s",st);
            }
        }
        return(0);
    }

    if(strcmp(argv[1],"-E")==0)
    {
        int i=2;
        int flag=0;
        while(argv[i]!=NULL)
        {
            flag=1;
            red_dol(org(argv[i]));
            i+=1;
        }
        if(flag==0)
        {
            while(1)
            {
                char *st;
                fgets(st,10000,stdin);
                printf("%s",st);
            }
        }
        return(0);
    }


    else if(strcmp(argv[1],"--help")==0)
    {
        printf("Usage: cat [OPTION]... [FILE]...\n");
        printf("Concatenate FILE(s) to standard output.\n\n");
        printf("With no FILE, or when FILE is -, read standard input.\n\n");
        printf("  -n,             number all output lines\n\n");
        printf("  -E,             display $ at end of each line\n");
        printf("         --help     display this help and exit\n\n");
        printf("cat f - g  Output f's contents, then standard input, then g's contents.\n");
        printf("cat        Copy standard input to standard output.");
        return(0);
    }

    else
    {
        char* tmp=argv[1];
        if (*tmp=='-')
        {
            printf("cat: invalid option -- '%s'\n",argv[1]);
            printf("Try 'cat --help' for more information.\n");
            return(-1);
        }
        int i=1;
        int flag=0;
        while(argv[i]!=NULL)
        {
            flag=1;
            red(org(argv[i]));
            i+=1;
        }
        if(flag==0)
        {
            while(1){}
        }
        return(0);
    }
}

rem.c
#include<stdio.h>
#include<unistd.h>
#include<string.h>
#include<sys/stat.h>
#include<sys/types.h>

char *org(char *name)
{
    char* tmp=name;
    while (*tmp!='\0')
    {
        if(*tmp=='@')
            *tmp=' ';
        tmp++;
    }
    return name;
         
}
int main(int argc,char **argv)
{   
    if(strcmp(argv[1],"-d")==0)
    {
        int i=2;
        int flag=0;
        while(argv[i]!=NULL)
        {
            flag=1;
            if(rmdir(org(argv[i]))!=0)
                printf("rm: cannot remove directory ‘%s’: No such file or directory\n",org(argv[i]));
            i+=1;
        }
        if(flag==0)
        {
            printf("rm: missing operand\n");
            printf("Try 'rm --help' for more information.\n");
        }
        return(0);
    }

    else if(strcmp(argv[1],"--help")==0)
    {
        printf("Usage: rm [OPTION]... [FILE]...\n");
        printf("Remove (unlink) the FILE(s).\n\n");
        printf("Create the DIRECTORY(ies), if they do not already exist.\n\n");
        printf("  -d,             remove empty directories\n");
        printf("--help        display this help and exit\n\n");
        printf("By default, rm does not remove directories.\n");
        return(0);
    }

    else
    {
        char* tmp=argv[1];
        if (*tmp=='-')
        {
            printf("rm: invalid option -- '%s'\n",argv[1]);
            printf("Try 'rm --help' for more information.\n");
            return(-1);
        }
        
        
        int i=1;
        int flag=0;
        while(argv[i]!=NULL)
        {
            flag=1;
            if(remove(org(argv[i]))!=0)
                printf("rm: cannot remove '%s': No such file or directory\n",org(argv[i]));
            i+=1;
        }
        if(flag==0)
        {
            printf("rm: missing operand\n");
            printf("Try 'rm --help' for more information.\n");
        }
        return(0);
    }
}

date.c
#include<stdio.h>
#include<time.h>
#include<string.h>

int main(int argc,char **argv)
{
    int i=0;
    if( argc >= 2 )
    {
      while(argv[i]!=NULL)
        i+=1;
    }
    int hr,min,sec,dat,mon,yr;
    char* month;
    time_t now;
    time(&now);
    struct tm *tt;
    if(i==2 && strcmp(argv[1],"-R")!=0)
        tt=gmtime(&now);
    else
        tt=localtime(&now);
    hr = tt->tm_hour;	  	
	min = tt->tm_min;
	sec = tt->tm_sec;	
	dat = tt->tm_mday;
    mon = tt->tm_mon + 1; 	  	
	yr = tt->tm_year + 1900;


    char* day=strtok(ctime(&now)," ");
    char* nday;
    char* nmon;
    if (strcmp(day,"Mon")==0)
        nday="Monday";
    else if(strcmp(day,"Tue")==0)
        nday="Tuesday";
    else if(strcmp(day,"Wed")==0)
        nday="Wednesday";
    else if(strcmp(day,"Thu")==0)
        nday="Thursday";
    else if(strcmp(day,"Fri")==0)
        nday="Friday";
    else if(strcmp(day,"Sat")==0)
        nday="Saturday";
    else
        nday="Sunday"; 


    if(mon==1)
    {
        month="January";
        nmon="Jan";
    }
    else if(mon==2)
    {
        month="Febraury";
        nmon="Feb";
    }
    else if(mon==3)
    {
        nmon="Mar";
        month="March";
    }
    else if(mon==4)
    {
        nmon="Apr";
        month="April";
    }
    else if(mon==5)
    {
        nmon="May";
        month="May";
    }
    else if(mon==6)
    {
        nmon="Jun";
        month="June";
    }
    else if(mon==7)
    {
        nmon="Jul";
        month="July";
    }
    else if(mon==8)
    {
        month="August";
        nmon="Aug";
    }
    else if(mon==9)
    {
        nmon="Sep";
        month="September";
    }
    else if(mon==10)
    {
        nmon="Oct";
        month="October";
    }
    else if(mon==11)
    {
        nmon="Nov";
        month="November";
    }
    else
    {   
        nmon="Dec";
        month="December";
    }
    
    
    if(i>=2)
    {
        if(i>2)
        {
            printf("date: extra operand's\n");
            return(-1);
        }
        if(strcmp(argv[1],"-R")==0)
        {
            printf("%s, %d %s %d %02d:%02d:%02d +0530\n",day,dat,nmon,yr,hr,min,sec);
            return(0);
        }
        else if(strcmp(argv[1],"-u")==0)
        {
            if(hr<=12)
                printf("%s %d %s %d %02d:%02d:%02d AM UTC\n",nday,dat,month,yr,hr,min,sec);
            else
                printf("%s %d %s %d %02d:%02d:%02d PM UTC\n",nday,dat,month,yr,hr-12,min,sec);
            return(0);
        }
        else
        {
            printf("date: invalid date\n");
            return(-1);
        }
    }
    else
    {   
        if(hr<=12)
            printf("%s %d %s %d %02d:%02d:%02d AM IST\n",nday,dat,month,yr,hr,min,sec);
        else
            printf("%s %d %s %d %02d:%02d:%02d PM IST\n",nday,dat,month,yr,hr-12,min,sec);
        return(0);
    }
}

mkdr.c
#include<stdio.h>
#include<unistd.h>
#include<string.h>
#include<sys/stat.h>
#include<sys/types.h>

char *org(char *name)
{
    char* tmp=name;
    while (*tmp!='\0')
    {
        if(*tmp=='@')
            *tmp=' ';
        tmp++;
    }
    return name;
         
}
int main(int argc,char **argv)
{   
    if(argv[1]==NULL)
    {
        printf("mkdir: missing operand\n");
        printf("Try 'mkdir --help' for more information.\n");
        return(0);
    }
    if(strcmp(argv[1],"-v")==0)
    {
        int i=2;
        int flag=0;
        while(argv[i]!=NULL)
        {
            flag=1;
            if(mkdir(org(argv[i]),S_IRWXU|S_IRWXG|S_IRWXO)!=0)
                printf("mkdir: cannot create directory ‘%s’: File exists\n",org(argv[i]));
            else
                printf("mkdir: created directory '%s'\n",org(argv[i]));
            i+=1;
        }
        if(flag==0)
        {
            printf("rm: missing operand\n");
            printf("Try 'mkdir --help' for more information.\n");
        }
        return(0);
    }

    else if(strcmp(argv[1],"--help")==0)
    {
        printf("Usage: mkdir [OPTION]... DIRECTORY...\n");
        printf("Create the DIRECTORY(ies), if they do not already exist.\n\n");
        printf("  -v,             print a message for each created directory\n\n");
        printf("         --help     display this help and exit\n");
        return(0);
    }

    else
    {
        char* tmp=argv[1];
        if (*tmp=='-')
        {
            printf("mkdir: invalid option -- '%s'\n",argv[1]);
            printf("Try 'mkdir --help' for more information.\n");
            return(-1);
        }
        int i=1;
        int flag=0;
        while(argv[i]!=NULL)
        {
            flag=1;
            if(mkdir(org(argv[i]),S_IRWXU|S_IRWXG|S_IRWXO)!=0)
                printf("mkdir: cannot create directory ‘%s’: File exists\n",org(argv[i]));
            i+=1;
        }
        if(flag==0)
        {
            printf("rm: missing operand\n");
            printf("Try 'mkdir --help' for more information.\n");
        }
        return(0);
    }
}

2019260_Ass1_2.c
#include<stdio.h>
#include<sys/types.h>
#include<unistd.h> 
#include<sys/wait.h>
#include<string.h>
#include<stdlib.h>
char cur_path[1000];

#define clear() printf("\033[H\033[J")
size_t bufsize = 64;
char *command;
char **command_split;
char cur_his[100000];



char **split(char *command)
{
    char **args=malloc(bufsize*sizeof(char*));
    char *token;
    if (!args) {
    fprintf(stderr, "lsh: allocation error\n");
    exit(EXIT_FAILURE);
    }
    int i=0;
    token=strtok(command," ");
    do
    {
        args[i]=token;
        i+=1;
        if (i>=bufsize)
        {
            args=realloc(args,(2*bufsize)*sizeof(char*));
        }
        token = strtok(NULL, " ");
        
    } while (token!=NULL);
    args[i]=NULL;
    return args;
}

int command_size(char **command_split)
{
    int i=0;
    while (command_split[i]!=NULL)
        i++;
    return i;
}

char *org(char *name)
{
    char* tmp=name;
    while (*tmp!='\0')
    {
        if(*tmp=='@')
            *tmp=' ';
        tmp++;
    }
    return name;
         
}

int ext_date(char **command_split)
{
    char pp[1000];
    pp[0]='\0';
    strcat(pp,cur_path);
    strcat(pp,"/");
    strcat(pp,"./date");
    pid_t pid;
    pid=fork();
    int status;
    if(pid<0)
    {
        fprintf(stderr,"Fork Failed");
        return 1;
    }
    else if(pid==0)
        execvp(pp,command_split);
    else
        waitpid(pid,&status,0);
}

int remov(char **command_split)
{
    char pp[1000];
    pp[0]='\0';
    strcat(pp,cur_path);
    strcat(pp,"/");
    strcat(pp,"./rem");
    pid_t pid;
    pid=fork();
    int status;
    if(pid<0)
    {
        fprintf(stderr,"Fork Failed");
        return 1;
    }
    else if(pid==0)
        execvp(pp,command_split);
    else
        waitpid(pid,&status,0);
}

int mk(char **command_split)
{
    char pp[1000];
    pp[0]='\0';
    strcat(pp,cur_path);
    strcat(pp,"/");
    strcat(pp,"./mkdr");
    pid_t pid;
    pid=fork();
    int status;
    if(pid<0)
    {
        fprintf(stderr,"Fork Failed");
        return 1;
    }
    else if(pid==0)
        execvp(pp,command_split);
    else
        waitpid(pid,&status,0);
}

int ct(char **command_split)
{
    char pp[1000];
    pp[0]='\0';
    strcat(pp,cur_path);
    strcat(pp,"/");
    strcat(pp,"./cat");
    pid_t pid;
    pid=fork();
    int status;
    if(pid<0)
    {
        fprintf(stderr,"Fork Failed");
        return 1;
    }
    else if(pid==0)
        execvp(pp,command_split);
    else
        waitpid(pid,&status,0);
}

int ls(char **command_split)
{
    char pp[1000];
    pp[0]='\0';
    strcat(pp,cur_path);
    strcat(pp,"/");
    strcat(pp,"./ls");
    pid_t pid;
    pid=fork();
    int status;
    if(pid<0)
    {
        fprintf(stderr,"Fork Failed");
        return 1;
    }
    else if(pid==0)
        execvp(pp,command_split);
    else
        waitpid(pid,&status,0);
}

int imp_echo(char **command_split)
{
    int sl=command_size(command_split);
    if(sl==1)
    {
        printf("\n");
        return(0);
    }
    else if(sl==2)
    {
        printf("%s\n",org(command_split[1]));
        return(0);
    }
    else
    {
        if(strcmp(command_split[1],"-n")==0)
        {
            printf("%s",org(command_split[2]));
            return(0);
        }
        if(strcmp(command_split[1],"-E")==0)
        {
            printf("%s",org(command_split[2]));
            return(0);
        }
        else
        {
            int i=1;
            while (command_split[i]!=NULL)
            {
                printf("%s ",command_split[i]);
                i++;
            }
            printf("\n");
            return(0);
        }
        
    }
}

int imp_his(char **command_split)
{
    int sl=command_size(command_split);
    if(sl==1)
    {
        int count=1;
        printf("%d  ",count);
        for(int i=0;cur_his[i]!='\0';i++)
        {
            if(cur_his[i]=='-'&&cur_his[i+1]!='\0')
            {
                printf("\n");
                count+=1;
                printf("%d  ",count);
            }
            else
                printf("%c",cur_his[i]);
        }
        printf("\n");
        return(0);
    }
    if(sl>=2 && strcmp(command_split[1],"-w")==0)
    {
        FILE *file_ptr;
        file_ptr=fopen("history.txt","a");
        for(int i=0;cur_his[i]!='\0';i++)
        {
            if(cur_his[i]=='-'&&cur_his[i+1]!='\0')
                fputc('\n',file_ptr);
            else 
                fputc(cur_his[i],file_ptr);
        }
        return(0);
    }
    else if(sl>=2 && strcmp(command_split[1],"--help")==0) 
    {
        printf("  Display or manipulate the history list.\n\n");
        printf("  Display the history list with line numbers, prefixing each modified\n");
        printf("  entry with a `*'.  An argument of N lists only the last N entries.\n\n");
        printf("  Options:\n\n");
        printf("  -w        write the current history to the history file\n\n");
        printf("Exit Status:\n");
        printf("  Returns success unless an invalid option is given or an error occurs.\n");
        return(0);
    }
    else
    {
        printf("bash: history: '%s': numeric argument required\n",command_split[1]);
        return(-1);
    }
}

int imp_cd(char **command_split)
{
    int sl=command_size(command_split);
    if(sl==1)
    {
        chdir(getenv("HOME"));
        return(0);
    }
    if(sl==2)
    {
        if(chdir(org(command_split[1]))!=0)
            printf("bash: cd: '%s': No such file or directory\n",command_split[1]);
        return(0);
    }
    if(sl>2)
    {
        if(strcmp(org(command_split[1]),"-e")!=0 && strcmp(command_split[1],"--help")!=0)
        {
            printf("bash: cd: too many arguments\n");
            return(0);
        }
        else if(strcmp(command_split[1],"-e")==0)
        {
            if(chdir(org(command_split[1]))!=0)
            {
                printf("bash: cd: '%s': No such file or directory\n",command_split[1]);
                return(-1);
            }
            return(0);
        }
        else if(strcmp(command_split[1],"--help")==0)
        {
            printf("cd: cd [-P] [dir]\n");
            printf("Change the shell working directory.\n\n");
            printf("Change the current directory to DIR.  The default DIR is the value of the HOME shell variable.\n\n");
            printf("Options:\n");
            printf("-P        use the physical directory structure without following symbolic links: resolve symbolic links in DIR before processing instances of `..'\n");
            printf("--help    display this help and exit\n");
            return(0);
        }
    }
}

int imp_pwd(char **command_split)
{
    int sl=command_size(command_split);
    if(sl==1)
    {
        char s[1000];
        printf("%s\n",getcwd(s,1000));
        return(0);
    }
    if(sl>=1)
    {
        if(strcmp(command_split[1],"-L")==0)
        {
            char s[1000];
            printf("%s\n",getcwd(s,1000));
            return(0);
        }
        else if(strcmp(command_split[1],"--help")==0)
        {
            printf("pwd: pwd [-L]\n");
            printf("Print the name of the current working directory.\n\n");
            printf("Options:\n");
            printf("-P        print the physical directory, without any symbolic links\n");
            printf("--help    display this help and exit\n");
            return(0);
        }
        else
        {
            printf("bash: pwd: -%s: invalid option\n",command_split[1]);
            printf("pwd: usage: pwd [-L]\n");
            return(0);
        }
    }
}

int main()
{
    clear();
    getcwd(cur_path, 1000);
    while (1)   
    {
        char pp[100];
        int com_size;
        printf("[user@linux ~]:%s$ ",getcwd(pp,sizeof(pp)));
        command=(char *) malloc(bufsize*sizeof(char));
        if( command == NULL)
        {
            perror("Unable to allocate buffer");
            exit(1);
        }
        getline(&command,&bufsize,stdin);
        int i;
        for(i=0;*(command+i)!='\n';++i);
        *(command+i)='\0';
        command_split=split(command);
        if (strcmp(command,"exit")==0)
            exit(EXIT_SUCCESS);
        else if(strcmp(command_split[0],"date")==0)
            ext_date(command_split);
        else if(strcmp(command_split[0],"rm")==0)
            remov(command_split);
        else if(strcmp(command_split[0],"mkdir")==0)
            mk(command_split);
        else if(strcmp(command_split[0],"cat")==0)
            ct(command_split);
        else if(strcmp(command_split[0],"ls")==0)
            ls(command_split);
        else if(strcmp(command_split[0],"echo")==0)
            imp_echo(command_split);
        else if(strcmp(command_split[0],"history")==0)
            imp_his(command_split);
        else if(strcmp(command_split[0],"cd")==0)
            imp_cd(command_split);
        else if(strcmp(command_split[0],"pwd")==0)
            imp_pwd(command_split);
        else
            printf("bash: '%s': command not found\n",command_split[0]);

        strcat(command_split[0],"-");
        strcat(cur_his,command_split[0]);
        free(command);
        free(command_split);
    }
    strcpy(cur_his,"");
    return(0);
}

BashCode.c
#include<stdio.h>
#include<sys/types.h>
#include<unistd.h> 
#include<sys/wait.h>
#include<string.h>
#include<stdlib.h>
char cur_path[1000];

#define clear() printf("\033[H\033[J")
size_t bufsize = 64;
char *command;
char **command_split;
char cur_his[100000];



char **split(char *command)
{
    char **args=malloc(bufsize*sizeof(char*));
    char *token;
    if (!args) {
    fprintf(stderr, "lsh: allocation error\n");
    exit(EXIT_FAILURE);
    }
    int i=0;
    token=strtok(command," ");
    do
    {
        args[i]=token;
        i+=1;
        if (i>=bufsize)
        {
            args=realloc(args,(2*bufsize)*sizeof(char*));
        }
        token = strtok(NULL, " ");
        
    } while (token!=NULL);
    args[i]=NULL;
    return args;
}

int command_size(char **command_split)
{
    int i=0;
    while (command_split[i]!=NULL)
        i++;
    return i;
}

char *org(char *name)
{
    char* tmp=name;
    while (*tmp!='\0')
    {
        if(*tmp=='@')
            *tmp=' ';
        tmp++;
    }
    return name;
         
}

int ext_date(char **command_split)
{
    char pp[1000];
    pp[0]='\0';
    strcat(pp,cur_path);
    strcat(pp,"/");
    strcat(pp,"./date");
    pid_t pid;
    pid=fork();
    int status;
    if(pid<0)
    {
        fprintf(stderr,"Fork Failed");
        return 1;
    }
    else if(pid==0)
        execvp(pp,command_split);
    else
        waitpid(pid,&status,0);
}

int remov(char **command_split)
{
    char pp[1000];
    pp[0]='\0';
    strcat(pp,cur_path);
    strcat(pp,"/");
    strcat(pp,"./rem");
    pid_t pid;
    pid=fork();
    int status;
    if(pid<0)
    {
        fprintf(stderr,"Fork Failed");
        return 1;
    }
    else if(pid==0)
        execvp(pp,command_split);
    else
        waitpid(pid,&status,0);
}

int mk(char **command_split)
{
    char pp[1000];
    pp[0]='\0';
    strcat(pp,cur_path);
    strcat(pp,"/");
    strcat(pp,"./mkdr");
    pid_t pid;
    pid=fork();
    int status;
    if(pid<0)
    {
        fprintf(stderr,"Fork Failed");
        return 1;
    }
    else if(pid==0)
        execvp(pp,command_split);
    else
        waitpid(pid,&status,0);
}

int ct(char **command_split)
{
    char pp[1000];
    pp[0]='\0';
    strcat(pp,cur_path);
    strcat(pp,"/");
    strcat(pp,"./cat");
    pid_t pid;
    pid=fork();
    int status;
    if(pid<0)
    {
        fprintf(stderr,"Fork Failed");
        return 1;
    }
    else if(pid==0)
        execvp(pp,command_split);
    else
        waitpid(pid,&status,0);
}

int ls(char **command_split)
{
    char pp[1000];
    pp[0]='\0';
    strcat(pp,cur_path);
    strcat(pp,"/");
    strcat(pp,"./ls");
    pid_t pid;
    pid=fork();
    int status;
    if(pid<0)
    {
        fprintf(stderr,"Fork Failed");
        return 1;
    }
    else if(pid==0)
        execvp(pp,command_split);
    else
        waitpid(pid,&status,0);
}

int imp_echo(char **command_split)
{
    int sl=command_size(command_split);
    if(sl==1)
    {
        printf("\n");
        return(0);
    }
    else if(sl==2)
    {
        printf("%s\n",org(command_split[1]));
        return(0);
    }
    else
    {
        if(strcmp(command_split[1],"-n")==0)
        {
            printf("%s",org(command_split[2]));
            return(0);
        }
        if(strcmp(command_split[1],"-E")==0)
        {
            printf("%s",org(command_split[2]));
            return(0);
        }
        else
        {
            int i=1;
            while (command_split[i]!=NULL)
            {
                printf("%s ",command_split[i]);
                i++;
            }
            printf("\n");
            return(0);
        }
        
    }
}

int imp_his(char **command_split)
{
    int sl=command_size(command_split);
    if(sl==1)
    {
        int count=1;
        printf("%d  ",count);
        for(int i=0;cur_his[i]!='\0';i++)
        {
            if(cur_his[i]=='-'&&cur_his[i+1]!='\0')
            {
                printf("\n");
                count+=1;
                printf("%d  ",count);
            }
            else
                printf("%c",cur_his[i]);
        }
        printf("\n");
        return(0);
    }
    if(sl>=2 && strcmp(command_split[1],"-w")==0)
    {
        FILE *file_ptr;
        file_ptr=fopen("history.txt","a");
        for(int i=0;cur_his[i]!='\0';i++)
        {
            if(cur_his[i]=='-'&&cur_his[i+1]!='\0')
                fputc('\n',file_ptr);
            else 
                fputc(cur_his[i],file_ptr);
        }
        return(0);
    }
    else if(sl>=2 && strcmp(command_split[1],"--help")==0) 
    {
        printf("  Display or manipulate the history list.\n\n");
        printf("  Display the history list with line numbers, prefixing each modified\n");
        printf("  entry with a `*'.  An argument of N lists only the last N entries.\n\n");
        printf("  Options:\n\n");
        printf("  -w        write the current history to the history file\n\n");
        printf("Exit Status:\n");
        printf("  Returns success unless an invalid option is given or an error occurs.\n");
        return(0);
    }
    else
    {
        printf("bash: history: '%s': numeric argument required\n",command_split[1]);
        return(-1);
    }
}

int imp_cd(char **command_split)
{
    int sl=command_size(command_split);
    if(sl==1)
    {
        chdir(getenv("HOME"));
        return(0);
    }
    if(sl==2)
    {
        if(chdir(org(command_split[1]))!=0)
            printf("bash: cd: '%s': No such file or directory\n",command_split[1]);
        return(0);
    }
    if(sl>2)
    {
        if(strcmp(org(command_split[1]),"-e")!=0 && strcmp(command_split[1],"--help")!=0)
        {
            printf("bash: cd: too many arguments\n");
            return(0);
        }
        else if(strcmp(command_split[1],"-e")==0)
        {
            if(chdir(org(command_split[1]))!=0)
            {
                printf("bash: cd: '%s': No such file or directory\n",command_split[1]);
                return(-1);
            }
            return(0);
        }
        else if(strcmp(command_split[1],"--help")==0)
        {
            printf("cd: cd [-P] [dir]\n");
            printf("Change the shell working directory.\n\n");
            printf("Change the current directory to DIR.  The default DIR is the value of the HOME shell variable.\n\n");
            printf("Options:\n");
            printf("-P        use the physical directory structure without following symbolic links: resolve symbolic links in DIR before processing instances of `..'\n");
            printf("--help    display this help and exit\n");
            return(0);
        }
    }
}

int imp_pwd(char **command_split)
{
    int sl=command_size(command_split);
    if(sl==1)
    {
        char s[1000];
        printf("%s\n",getcwd(s,1000));
        return(0);
    }
    if(sl>=1)
    {
        if(strcmp(command_split[1],"-L")==0)
        {
            char s[1000];
            printf("%s\n",getcwd(s,1000));
            return(0);
        }
        else if(strcmp(command_split[1],"--help")==0)
        {
            printf("pwd: pwd [-L]\n");
            printf("Print the name of the current working directory.\n\n");
            printf("Options:\n");
            printf("-P        print the physical directory, without any symbolic links\n");
            printf("--help    display this help and exit\n");
            return(0);
        }
        else
        {
            printf("bash: pwd: -%s: invalid option\n",command_split[1]);
            printf("pwd: usage: pwd [-L]\n");
            return(0);
        }
    }
}

int main()
{
    clear();
    getcwd(cur_path, 1000);
    while (1)   
    {
        char pp[100];
        int com_size;
        printf("[user@linux ~]:%s$ ",getcwd(pp,sizeof(pp)));
        command=(char *) malloc(bufsize*sizeof(char));
        if( command == NULL)
        {
            perror("Unable to allocate buffer");
            exit(1);
        }
        getline(&command,&bufsize,stdin);
        int i;
        for(i=0;*(command+i)!='\n';++i);
        *(command+i)='\0';
        command_split=split(command);
        if (strcmp(command,"exit")==0)
            exit(EXIT_SUCCESS);
        else if(strcmp(command_split[0],"date")==0)
            ext_date(command_split);
        else if(strcmp(command_split[0],"rm")==0)
            remov(command_split);
        else if(strcmp(command_split[0],"mkdir")==0)
            mk(command_split);
        else if(strcmp(command_split[0],"cat")==0)
            ct(command_split);
        else if(strcmp(command_split[0],"ls")==0)
            ls(command_split);
        else if(strcmp(command_split[0],"echo")==0)
            imp_echo(command_split);
        else if(strcmp(command_split[0],"history")==0)
            imp_his(command_split);
        else if(strcmp(command_split[0],"cd")==0)
            imp_cd(command_split);
        else if(strcmp(command_split[0],"pwd")==0)
            imp_pwd(command_split);
        else
            printf("bash: '%s': command not found\n",command_split[0]);

        strcat(command_split[0],"-");
        strcat(cur_his,command_split[0]);
        free(command);
        free(command_split);
    }
    strcpy(cur_his,"");
    return(0);
}
